src/data/set/ncard.lean

```diff
@@ -0,0 +1,694 @@
+/-
+Copyright (c) 2023 Peter Nelson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Peter Nelson
+-/
+import data.finite.card
+
+/-!
+# Noncomputable Set Cardinality
+
+We define the cardinality `set.ncard s` of a set `s` as a natural number. This function is
+noncomputable (being defined in terms of `nat.card`) and takes the value `0` if `s` is infinite.
+
+This can be seen as an API for `nat.card α` in the special case where `α` is a subtype arising from
+a set. It is intended as an alternative to `finset.card` and `fintype.card`,  both of which contain
+data in their definition that can cause awkwardness when using `set.to_finset`.  Using `set.ncard`
+allows cardinality computations to avoid `finset`/`fintype` completely, staying in `set` and letting
+finiteness be handled explicitly, or (where a `finite α` instance is present and the sets are
+in `set α`) via `auto_param`s.
+
+## Main Definitions
+
+* `set.ncard s` is the cardinality of the set `s` as a natural number, provided `s` is finite.
+  If `s` is infinite, then `set.ncard s = 0`.
+* `to_finite_tac` is a tactic that tries to synthesize an `set.finite s` argument with
+  `set.to_finite`. This will work for `s : set α` where there is a `finite α` instance.
+
+## Implementation Notes
+
+The lemmas in this file are very similar to those in `data.finset.card`, but with `set` operations
+instead of `finset`; most of the proofs invoke their `finset` analogues. Nearly all the lemmas
+require finiteness of one or more of their arguments. We provide this assumption with an
+`auto_param` argument of the form `(hs : s.finite . to_finite_tac)`, where `to_finite_tac` will find
+a `finite s` term in the cases where `s` is a set in a `finite` type.
+
+Often, where there are two set arguments `s` and `t`, the finiteness of one follows from the other
+in the context of the lemma, in which case we only include the ones that are needed, and derive the
+other inside the proof. A few of the lemmas, such as `ncard_union_le` do not require finiteness
+arguments; they are are true by coincidence due to junk values.
+-/
+
+open_locale classical
+
+variables {α β : Type*} {s t : set α} {a b x y : α} {f : α → β}
+
+namespace set
+
+
+/-- The cardinality of `s : set α` . Has the junk value `0` if `s` is infinite -/
+noncomputable def ncard (s : set α) := nat.card s
+
+/-- A tactic that finds a `t.finite` term for a set `t` in a `finite` type. -/
+meta def to_finite_tac : tactic unit := `[try {apply set.to_finite}]
+
+lemma ncard_def (s : set α) : s.ncard = nat.card s := rfl
+
+lemma ncard_eq_to_finset_card (s : set α) (hs : s.finite . to_finite_tac) :
+  s.ncard = finset.card (@set.to_finset α s (@fintype.of_finite _ hs.to_subtype)) :=
```

Comment:
```suggestion
  s.ncard = hs.to_finset.card :=
```
feels more idiomatic

Reply:
Thanks :)

Reply:
And thanks again. This changed really cleaned up a lot of the proofs.